keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,241 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+## Summary
+
+This KEP tries to integrate implementation of volume limits in scheduler for in-tree volumes (AWS EBS, GCE PD, AZURE DD, OpenStack Cinder) and CSI into one predicate. As result, in-tree volumes and corresponsing CSI driver can share the same volume limit. 
+
+## Motivation
+
+Current scheduler predicates for scheduling of pods with volumes is based on `node.status.capacity` and `node.status.allocatable`. It works well for hardcoded predicates for volume limits on AWS (`MaxEBSVolumeCount`), GCE(`MaxGCEPDVolumeCount`), Azure (`MaxAzureDiskVolumeCount`) and OpenStack (`MaxCinderVolumeCount`).
+
+It is problematic for CSI (`MaxCSIVolumeCountPred`):
+
+- `ResourceName` is limited to 63 characters. We must prefix `ResourceName` with unique string (such as `attachable-volumes-csi-<driver name>`) so it cannot collide with existing resources like `cpu` or `memory`. But `<driver name>` itself is up to 63 character long, so we ended up with using SHA-sums of driver name to keep the `ResourceName` unique, which is not user readable.
+- CSI driver cannot share its limits with in-tree volume plugin e.g. when running pods with AWS EBS in-tree volumes and `ebs.csi.aws.com` CSI driver on the same node.
+- `node.status` size increases with each installed CSI driver. Node objects is big enough already.
+
+### Goals
+
+- User can run use PVs both with in-tree volume plugins and CSI and they will share their limits. There is only one scheduler predicate that handles both kind of volumes.
+
+- Existing predicates for in-tree volumes `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` are removed (with deprecation period).
+  - When both deprecated in-tree predicate and CSI predicate are enabled, only one of them does useful work and the other is NOOP to save CPU.
+
+- CPU consumption of the scheduler does not rise, i.e. test XYZ passes. (TODO: fill test name)
+
+### Non-Goals
+
+
+## Proposal
+
+* Track volume limits for both in-tree volume plugins and CSI drivers in `CSINode` objects instead of `Node`.
+  * `Node` object is already big enough.
+  * Get rid of prefix + SHA for `ResourceName` of CSI volumes.
+  * In-tree volume plugin can share limits with CSI driver that uses the same storage backend.
+
+* Kubelet will create `CSINode` instance during initial node registration together with `Node` object.
+  * Limits of each in-tree volume plugin will be added to `CSINode.status.capacity` and `CSINode.status.allocatable`.
```

Comment:
Since `capacity` could have different meaning for storage, my vote is to keep this field name `attach_limit`. So something like `CSINode.Status.AttachLimit` . I am also not sure if we want to have both capacity and allocatable here.

Reply:
>  I am also not sure if we want to have both capacity and allocatable here.

@gnufied is there any reason why we only need one of `capacity` or `allocatable`? From the `Node.status` [definition](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L4026), `capacity` means total resource count while `allocatable` means available resource count. Should we keep both to be able to surface the same amount of information about attachment ?

Reply:
> What about AttachVolumeLimit, like the feature gate?

As I wrote above, this feature is not about attach limits. You can limit non-attachable volumes too.

>  I am also not sure if we want to have both capacity and allocatable here.

On second thoughts, I think we should keep `capacity` (= what's reported by volume plugin / CSI driver) and have a way for users to reduce `allocatable` (= what's available to pods), like with other resources. Adding `--system-reserved-volumes=` and `--kube-reserved-volumes=` to kubelet could be straightforward. I'd love to have some smart autodetect that would e.g. reduce `allocatable` by nr. of volumes that are attached/mounted outside of /var/lib/kubelet, but CSI does not allow us to list these "system" volumes.

Reply:
Talked to @gnufied, let's skip `capacity` and `--system-reserved-volumes` for now. We can always add it later.